### PR TITLE
New Hopper blockbase

### LINKED ISSUE #62: Hopper Mod Element
A mod element option for a hopper. It would allow interaction with other hoppers and TileEntities, but can be disabled as an option. You can adjust the number of slots and the stack sizes for each slot, and maybe even set it up to use FE and Fluids (basically, easy to make pipes).
Another option can disable use in certain directions, like disable going down, but enable going up, east, west, north, south; or disable North and South, but go the other directions.
There would be different block models (optional), depending on which directions you enable and disable.
You can adjust the speed at which items are transferred, and maybe filter options.
That's about all I got right now, I'll expand on this later as I'm short for time.

This could also just be an option for blocks, but with the different models, I don't think it could be.

### PR DESCRIPTION
**Pull request by Max094Reikeb**
This PR adds a new blockbases: Hopper
Closes #62 

- [x] Add Hopper blockbase
- [x] Block "titleentity" option if hopper blockbase is selected
- [ ] Add ability to send items to other hoppers
- [ ] Fix item-only issue

### CODE REVIEW

**Review by Goldorion — COMMENTED** [review]
Why do you add new procedure blocks?

**Comment by Max094Reikeb** [comment]
> Why do you add new procedure blocks?

I don't? Where do I add procedure blocks?

**Review by Goldorion — CHANGES_REQUESTED** [review]
You forgot the code in the two code template files. You have to add a line for the "extends ..." and another one for the "super(..." (the content of the "super" depends on the block base).

**Review by Goldorion — CHANGES_REQUESTED** [review]
You forgot the lines to extend the Hopper class.

**Comment by Max094Reikeb** [comment]
> You forgot the lines to extend the Hopper class.

How do I do that?

File: plugins/generator-1.16.4/forge-1.16.4/templates/block.java.ftl [hunk 1]
```diff
@@ -425,7 +449,8 @@ public class ${name}Block extends ${JavaModName}Elements.ModElement {
         @Override public FluidState getFluidState(BlockState state) {
             return state.get(WATERLOGGED) ? Fluids.WATER.getStillFluidState(false) : super.getFluidState(state);
         }
-	
+		</#if>
```
Comment (Goldorion) [inline plugins/generator-1.16.4/forge-1.16.4/templates/block.java.ftl:452]: What's the purpose of this </#if>?
It seems this line cause a build error.

Comment (Max094Reikeb) [reply plugins/generator-1.16.4/forge-1.16.4/templates/block.java.ftl]: Fixed.

**Review by Goldorion — CHANGES_REQUESTED** [review]
I tested in-game (1.15.2 only), and I noticed a few things.
1. Custom hoppers can not send items to another hopper (the same custom hopper) or to a Vanilla hopper.
2. If we enable the tile entity with the Hopper block base, the block doesn't act as a hopper. So you should not enable this option when the Hopper block base is selected.
3. The item entity when dropped doesn't have the same size as other item entities. (Example on the screenshot below)
4. When the hopper contains an item, it can't contain another type of item (e.g. if it has a diamond, it won't accept a gold ingot). To accept another type of item, we need to replace the hopper.
5. When the player has the custom hopper in his hand, the 3D model is used instead of the item texture (Vanilla hopper doesn't use the 3D mode).


I put this PR as a draft, so it will give you the time to fix these problems.

**Pull request closed without merging**